src/storage/aggregator.cc

```diff
@@ -0,0 +1,327 @@
+            buffer->aggr_val_.vdouble = origin_val;
+            break;
+        }
+        default: {
+            PDLOG(WARNING, "unsupported data type");
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Aggregator::FlushAggrBuffer(const std::string& key, const AggrBuffer& buffer) {
+    std::string encoded_row;
+    std::string aggr_val;
+    switch (aggr_col_type_) {
+        case DataType::kSmallInt:
+        case DataType::kInt:
+        case DataType::kBigInt: {
+            int64_t tmp_val = buffer.aggr_val_.vlong;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(int64_t));
+            break;
+        }
+        case DataType::kFloat: {
+            float tmp_val = buffer.aggr_val_.vfloat;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(float));
+            break;
+        }
+        case DataType::kDouble: {
+            double tmp_val = buffer.aggr_val_.vdouble;
+            aggr_val.assign(reinterpret_cast<char*>(&tmp_val), sizeof(double));
+            break;
+        }
+        default: {
+            PDLOG(WARNING, "unsupported data type");
+            return false;
+        }
+    }
+
+    int str_length = key.size() + aggr_val.size();
+    uint32_t row_size = row_builder_.CalTotalLength(str_length);
+    encoded_row.resize(row_size);
+    row_builder_.SetBuffer(reinterpret_cast<int8_t*>(&(encoded_row[0])), row_size);
+    row_builder_.AppendString(key.c_str(), key.size());
+    row_builder_.AppendTimestamp(buffer.ts_begin_);
+    row_builder_.AppendTimestamp(buffer.ts_end_);
+    row_builder_.AppendInt32(buffer.aggr_cnt_);
+    row_builder_.AppendString(aggr_val.c_str(), aggr_val.size());
+    row_builder_.AppendInt64(buffer.binlog_offset_);
+
+    int64_t time = ::baidu::common::timer::get_micros() / 1000;
+    dimensions_.Mutable(0)->set_key(key);
+    bool ok = aggr_table_->Put(time, encoded_row, dimensions_);
+    if (!ok) {
+        LOG(ERROR) << "aggregator put failed";
+        return false;
+    }
+
+    // init the next window range
+    auto& latest_val = aggr_buffer_map_.at(key);
+    int64_t latest_ts = latest_val.ts_end_ + 1;
+    latest_val.clear();
+    latest_val.ts_begin_ = latest_ts;
+    if (window_type_ == WindowType::kRowsRange) {
+        latest_val.ts_end_ = latest_ts + window_size_ - 1;
+    }
+
+    return true;
+}
+
+SumAggregator::SumAggregator(const ::openmldb::api::TableMeta& base_meta, const ::openmldb::api::TableMeta& aggr_meta,
+                             std::shared_ptr<Table> aggr_table, const uint32_t& index_pos, const std::string& aggr_col,
+                             const AggrType& aggr_type, const std::string& ts_col, WindowType window_tpye,
+                             uint32_t window_size)
+    : Aggregator(base_meta, aggr_meta, aggr_table, index_pos, aggr_col, aggr_type, ts_col, window_tpye, window_size) {}
+
+bool SumAggregator::UpdateAggrVal(codec::RowView* row_view, AggrBuffer* aggr_buffer) {
```

Comment:
row_view can be `const`?

Reply:
> row_view can be `const`?

call a non-const member function on a const object is not allowed.

Reply:
can mark these `GetValue` and other as `const`. They should be const functions.